zokrates_core/src/parser/parse/expression.rs

```diff
@@ -469,13 +472,12 @@ pub fn parse_expr<T: Field>(
             },
             Err(err) => Err(err),
         },
-        (Token::Ide(x), s1, p1) => match next_token::<T>(&s1, &p1) {
-            (Token::Open, s2, p2) => parse_function_call(x, s2, p2),
-            (Token::LeftBracket, s2, p2) => parse_array_select(x, s2, p2),
-            _ => match parse_term1(Expression::Identifier(x), s1, p1) {
-                Ok((e2, s2, p2)) => parse_expr1(e2, s2, p2),
-                Err(err) => Err(err),
+        (Token::Ide(x), s1, p1) => match parse_identified1(x, s1, p1) {
```

Comment:
@GuthL this is the main issue here imo, the one that cause the bug you witnessed: before this change in `parse_expr`, we only run `parse_term1`(which is "parse term starting with expression x") when x is an identifier. Otherwise (when x is a function call or an array access), it just returns x. Therefore we end up with `Sub(x, Sub(y, z)) == x - (y - z)` instead of the expected `Sub(Sub(x, y), z) == (x - y) - z`